### PR TITLE
Add broadcasting support for `tf.where`

### PR DESCRIPTION
**Pull request by yongtang**
This is a rework on #15982

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/ops/math_grad.py [hunk 1]
```diff
@@ -1308,6 +1308,38 @@ def _SelectGrad(op, grad):
...
+  gx = math_ops.reduce_sum(gx, keepdims=True, axis=axis)
+
+  gy = array_ops.where_v2(c, zeros, grad)
+  gy_shape = array_ops.shape(gy)
+  y_shape = array_ops.shape(y)
+  rankdiff_y = array_ops.rank(gy) - array_ops.rank(y)
+  # Reduce away broadcasted leading dims.
+  gy = math_ops.reduce_sum(gy, axis=math_ops.range(rankdiff_y))
+  # Reduce but keep y's 1-valued dims which were broadcast.
+  axis = array_ops.where(gy_shape[rankdiff_y:] > y_shape)
```
Comment (alextp) [inline tensorflow/python/ops/math_grad.py]: Why is this where and not where_v2?

**Review by alextp — CHANGES_REQUESTED** [review]
We also need to have tests for the gradient code. This means at least a gradient checker test for the cases with and without broadcast. Running all existing where gradient tests with where_v2 as well is also a good idea.

This is core TF functionality and accidentally breaking it would be very bad, so it's worth the extra effort to get right.

File: tensorflow/python/ops/math_grad.py (hunk 1 above)
Comment (yongtang) [reply tensorflow/python/ops/math_grad.py]: Thanks @alextp. The PR has been updated to use where_v2. In both cases (where_v2 and where), they behave the same though as `else` and `then` are `None` so follow the same rule. Updated to use `where_v2` for consistency.

**Comment by yongtang** [comment]
Thanks @alextp. The existing gradient tests also include where_v2 in `tensorflow/python/kernel_tests/cwise_ops_test.py`, the exiting non-gradient tests include where_v2 in   `tensorflow/python/kernel_tests/where_op_test.py`.

I will add additional test cases for broadcast that are special to where_v2 only. Will update again shortly.

**Review by alextp — DISMISSED** [review]

**Comment by yongtang** [comment]
@rthadur @alextp The failed tests are due to the api update. Should I update the api golden to make the tests pass?

**Comment by yongtang** [comment]
@alexp @rthadur I updated the PR for api goldens. Believe all tests should pass now. Please let me know if there are any issues.

File: tensorflow/tools/api/golden/v2/tensorflow.pbtxt [hunk 2]
```diff
@@ -904,6 +904,10 @@ tf_module {
     name: "searchsorted"
     argspec: "args=[\'sorted_sequence\', \'values\', \'side\', \'out_type\', \'name\'], varargs=None, keywords=None, defaults=[\'left\', \"<dtype: \'int32\'>\", \'None\'], "
   }
+  member_method {
+    name: "select_v2"
```
Comment (alextp) [inline tensorflow/tools/api/golden/v2/tensorflow.pbtxt]: The selectv2 op needs an apidef file in the python_api directory to mark it as hidden

**Review by alextp — DISMISSED** [review]
We also need to change the tf_upgrade_v2 script to rename tf.where_v2 to tf.compat.v2.where.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
It appears there is still something wrong:

```
import tensorflow.compat.v2 as tf
tf.enable_v2_behavior()

import numpy as np

print(np.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))
print(tf.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))

InvalidArgumentError: Inputs to operation Select of type Select must have the same size and shape.  Input 0: [1,1] != input 1: [1,10] [Op:Select]
```

**Comment by yongtang** [comment]
@martinwicke Sorry about that. I will take a look and will fix it as soon as possible.

**Comment by yongtang** [comment]
@martinwicke I tried with tf-nightly it looks fine:
```
# python
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np   
>>> 
>>> import tensorflow.compat.v2 as tf
>>> tf.enable_v2_behavior()
>>> 
>>> import numpy as np
>>> 
>>> print(np.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))
2019-05-17 16:20:24.989573: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2019-05-17 16:20:25.050076: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2500000000 Hz
2019-05-17 16:20:25.050408: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x563b3ef74c80 executing computations on platform Host. Devices:
2019-05-17 16:20:25.050437: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
>>> print(tf.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))
tf.Tensor([[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]], shape=(1, 10), dtype=float32)
>>> 
>>> print(tf.version.VERSION)
1.14.1-dev20190517
>>> print(tf.version.GIT_VERSION)
v1.12.1-2154-g3df6d99f3f
>>> 
```

Wondering if there is a version difference?

**Comment by martinwicke** [comment]
Yeah, you are right. For some reason my nightly didn't pick this up yet, and I'm confused about that.

**Comment by yongtang** [comment]
Thanks @martinwicke. Appreciate the help 👍

**Comment by rthadur** [comment]
Changes are merged in to Master, closing this PR.

**Pull request closed without merging**